pkg/appstate/daemonsets.go

```diff
@@ -0,0 +1,180 @@
+		},
+	}
+
+	informer := cache.NewSharedInformer(listwatch, &appsv1.DaemonSet{}, time.Minute)
+
+	eventHandler := &daemonSetEventHandler{
+		informers:       filterStatusInformersByResourceKind(informers, DaemonSetResourceKind),
+		resourceStateCh: resourceStateCh,
+		clientset:       clientset,
+		targetNamespace: targetNamespace,
+	}
+
+	runInformer(ctx, informer, eventHandler)
+}
+
+func (h *daemonSetEventHandler) ObjectCreated(obj interface{}) {
+	r := h.cast(obj)
+	if _, ok := h.getInformer(r); !ok {
+		return
+	}
+
+	h.resourceStateCh <- makeDaemonSetResourceState(r, h.calculateDaemonSetState(r))
+}
+
+func (h *daemonSetEventHandler) ObjectDeleted(obj interface{}) {
+	r := h.cast(obj)
+	if _, ok := h.getInformer(r); !ok {
+		return
+	}
+
+	h.resourceStateCh <- makeDaemonSetResourceState(r, types.StateMissing)
+}
+
+func (h *daemonSetEventHandler) ObjectUpdated(obj interface{}) {
+	r := h.cast(obj)
+	if _, ok := h.getInformer(r); !ok {
+		return
+	}
+
+	h.resourceStateCh <- makeDaemonSetResourceState(r, h.calculateDaemonSetState(r))
+}
+
+func (h *daemonSetEventHandler) getInformer(r *appsv1.DaemonSet) (types.StatusInformer, bool) {
+	if r != nil {
+		for _, informer := range h.informers {
+			if r.Namespace == informer.Namespace && r.Name == informer.Name {
+				return informer, true
+			}
+		}
+	}
+
+	return types.StatusInformer{}, false
+}
+
+func (h *daemonSetEventHandler) cast(obj interface{}) *appsv1.DaemonSet {
+	r, _ := obj.(*appsv1.DaemonSet)
+	return r
+}
+
+// The pods in a daemonset can be identified by the match label set in the daemonset and the
+// "controller-revision-hash" can be used to determine if they are all the in the same daemonset
+// version.
+func (h *daemonSetEventHandler) calculateDaemonSetState(r *appsv1.DaemonSet) types.State {
+	if r == nil {
+		return types.StateUnavailable
+	}
+
+	selector := ""
+	for key, value := range r.Spec.Selector.MatchLabels {
+		if len(selector) > 0 {
+			selector += ","
+		}
+		selector += fmt.Sprintf("%s=%s", key, value)
+	}
+
+	// Gather all pods by looping until server responds with an empty continue field.
+	pods := make([]corev1.Pod, 0)
+	for {
+		result, err := h.clientset.CoreV1().Pods(h.targetNamespace).List(context.TODO(), metav1.ListOptions{LabelSelector: selector})
+		if err != nil {
+			log.Printf("failed to get daemonset pod list: %s", err)
+			return types.StateUnavailable
+		}
+		pods = append(pods, result.Items...)
+
+		if len(result.Continue) == 0 {
+			break
+		}
+	}
+
+	// If the pod version labels are not all the same, then the daemonset is updating.
+	currentVersion := ""
+	for _, pod := range pods {
+		version, exists := pod.Labels[DaemonSetPodVersionLabel]
+		if !exists {
+			log.Printf("failed to find %s label for pod %s", DaemonSetPodVersionLabel, pod.Name)
+			return types.StateUnavailable
+		}
+
+		if len(currentVersion) == 0 {
+			currentVersion = version
+		} else {
+			if version != currentVersion {
```

Comment:
nit pick: can't this be combined with the `else` above to be `else if version != currentVersion`?